umpost/um2netcdf.py

```diff
@@ -92,46 +102,194 @@ def convert_proleptic(time):
+    dlon: (float) zonal spacing between longitude grid points. 
+    """
+    min_longitude = longitude_points[0]
+    min_longitude_u_ND = 0.5*dlon
+
+    return (
+        (min_longitude == 0 and grid_type == GRID_END_GAME) or
+        (
+            np.allclose(min_longitude_u_ND, min_longitude) and
+            grid_type == GRID_NEW_DYNAMICS
+        )
+    )
+
+
+def add_latlon_coord_bounds(cube_coordinate):
+    """
+    Add bounds to horizontal coordinate (longitude or latitude) if
+    they don't already exist. Edits coordinate in place.
+
+    Parameters
+    ----------
+    cube_coordinate: coordinate object from iris cube.
+    """
+    coordinate_name = cube_coordinate.name()
+    if coordinate_name not in [LON_COORD_NAME, LAT_COORD_NAME]:
+        raise ValueError(
+                f"Wrong coordinate {coordinate_name} supplied. "
+                f"Expected one of {LON_COORD_NAME}, {LAT_COORD_NAME}."
+            )
+
+    if cube_coordinate.has_bounds(): 
+        # Don't change bounds if already exist
+        return
+    
+    if len(cube_coordinate.points) > 1:
+        cube_coordinate.guess_bounds()
     else:
-        lon.var_name = 'lon'
+        # For length 1, assume it's global. guess_bounds doesn't work in this case
+        cube_coordinate.bounds = GLOBAL_COORD_BOUNDS[coordinate_name]
+
+
+
+def fix_latlon_coords(cube, grid_type, dlat, dlon):
+    """
+    Wrapper function to modify cube's horizontal coordinates 
+    (latitude and longitude). Converts to float64, adds grid bounds,
+    and renames coordinates. Modifies cube in place.
+
+    Parameters
+    ----------
+    cube: Iris cube object (modified in place).
+    grid_type: (string) model horizontal grid type.
+    dlat: (float) meridional spacing between latitude grid points.
+    NB - Only applies to variables on the main horizontal grids, 
+    and not the river grid.
+    dlon: (float) zonal spacing between longitude grid points. 
+    NB - Only applies to variables on the main horizontal grids, 
+    and not the river grid.
+    """
+
+    # TODO: Check that we don't need the double coordinate lookup.
+    try:
+        latitude_coordinate = cube.coord(LAT_COORD_NAME)
+        longitude_coordinate = cube.coord(LON_COORD_NAME)
+        # Force to double for consistency with CMOR
+        latitude_coordinate.points = latitude_coordinate.points.astype(np.float64)
+        longitude_coordinate.points = longitude_coordinate.points.astype(np.float64)
+        
+        add_latlon_coord_bounds(latitude_coordinate)
+        add_latlon_coord_bounds(longitude_coordinate)
+        
+        fix_lat_coord_name(latitude_coordinate, grid_type, dlat)
+        fix_lon_coord_name(longitude_coordinate, grid_type, dlon)
+        
+    except iris.exceptions.CoordinateNotFoundError:
+        warnings.warn(
+            "Missing latitude or longitude coordinate for variable (possible timeseries?): \n"
+            f"{cube}\n"
+        )
+        raise 
```

Comment:
There's a mixed error handling here with a warning & an exception. The original code shields the coord error behind an `Exception`, which is too generic.

Is suspected time series detection still considered a complete failure as `um2nc` doesn't handle time series?

To improve the error, how about:
```Python
except iris.exceptions.CoordinateNotFoundError as err:
    msg = "Missing latitude or longitude coordinate for variable (possible timeseries?): \n{cube}\n"
    raise PostProcessingError(msg) from err
```

Reply:
Good question. We expect the time series error to come up during automated post-processing of ESM1.5 simulations, as one of the model's output files contains time series. Whichever error is raised here, is the one that the conversion driver should allow to pass in:

https://github.com/ACCESS-NRI/um2nc-standalone/blob/1c86df606ca01eec677781dd2f99c10b26b07268/umpost/conversion_driver_esm1p5.py#L144-L150

I'm wondering whether it would make sense to add a new type of `PostProcessingError` here, e.g. `TimeSeriesError`, and then this would let us simplify the error handling in the driver script? Let me know what you think.

Reply:
If it's simpler to handle, adding a new error `UnsupportedTimeSeriesError" (?) is fine. Error handling needs rework module wide, so even if this is a temp solution, it can be refined later.

Reply:
Good idea, I've added in an `UndupportedTimeSeriesError`